injected/src/features/api-manipulation.js

```diff
@@ -0,0 +1,130 @@
+                if (!this.checkIsValidAPIChange(change)) {
+                    continue;
+                }
+                this.applyApiChange(scope, change);
+            }
+        }
+    }
+
+    /**
+     * Checks if the config API change is valid.
+     * @param {any} change
+     * @returns {change is APIChange}
+     */
+    checkIsValidAPIChange(change) {
+        if (typeof change !== 'object') {
+            return false;
+        }
+        if (change.type === 'remove') {
+            return true;
+        }
+        if (change.type === 'descriptor') {
+            if (change.enumerable && typeof change.enumerable !== 'boolean') {
+                return false;
+            }
+            if (change.configurable && typeof change.configurable !== 'boolean') {
+                return false;
+            }
+            return typeof change.getterValue !== 'undefined';
+        }
+        return false;
+    }
+
+    // TODO move this to schema definition imported from the privacy-config
+    /**
+     * @typedef {Object} APIChange
+     * @property {"remove"|"descriptor"} type
+     * @property {import('../utils.js').ConfigSetting} [getterValue] - The value returned from a getter.
+     * @property {boolean} [enumerable] - Whether the property is enumerable.
+     * @property {boolean} [configurable] - Whether the property is configurable.
+     */
+
+    /**
+     * Applies a change to DOM APIs.
+     * @param {string} scope
+     * @param {APIChange} change
+     * @returns {void}
+     */
+    applyApiChange(scope, change) {
+        const response = this.getGlobalObject(scope);
+        if (!response) {
+            return;
+        }
+        const [obj, key] = response;
+        if (change.type === 'remove') {
+            this.removeApiMethod(obj, key);
+        } else if (change.type === 'descriptor') {
+            this.wrapApiDescriptor(obj, key, change);
+        }
+    }
+
+    /**
+     * Removes a method from an API.
+     * @param {object} api
+     * @param {string} key
+     */
+    removeApiMethod(api, key) {
+        try {
+            if (key in api) {
+                delete api[key];
+            }
+        } catch (e) {}
+    }
+
+    /**
+     * Wraps a property with descriptor.
+     * @param {object} api
+     * @param {string} key
+     * @param {APIChange} change
+     */
+    wrapApiDescriptor(api, key, change) {
+        const getterValue = change.getterValue;
+        if (getterValue) {
+            this.wrapProperty(api, key, {
+                get: () => processAttr(getterValue, undefined),
+                enumerable: change.enumerable ?? true,
+                configurable: change.configurable ?? true,
+            });
+        }
+    }
+
+    /**
+     * Looks up a global object from a scope, e.g. 'Navigator.prototype'.
+     * @param {string} scope the scope of the object to get to.
+     * @returns {[object, string]|null} the object at the scope.
+     */
+    getGlobalObject(scope) {
```

Comment:
I could swear I wrote this function before, but maybe it was in previous life.
Maybe move this to wrapper-utils.js?

Reply:
Yeah we had one in runtime checks also. I did try looking if we had another.

Reply:
Branch: https://github.com/duckduckgo/content-scope-scripts/compare/jkt/apiManipulation...jkt/apiModificationUnitTestGetGlobalObject?expand=1